File: Assets/Scripts/Message.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
//using System.Text.Json;
using System;
using Newtonsoft.Json;
using System.Text;

public class Message
{
    public Dictionary<string, object> data;

    public Message(string action = "")
    {
        data = new Dictionary<string, object>();
        data["action"] = action;
    }

    public byte[] EncodeToSend(bool length)
    {
        string json_data = JsonConvert.SerializeObject(data);
        byte[] json_bytes = Encoding.UTF8.GetBytes(json_data);

        if (length)
        {
            byte[] length_bytes = BitConverter.GetBytes(json_bytes.Length);
            if (BitConverter.IsLittleEndian)
            {
                Array.Reverse(length_bytes);
            }
            return ConcatBytes(length_bytes, json_bytes);
        }
        else
        {
            return json_bytes;
        }
    }

    public void DecodeFromReceive(string convert)
    {
        data = JsonConvert.DeserializeObject<Dictionary<string, object>>(convert);
    }

    public void RequestRods()
    {
        Dictionary<string, object> rods = new Dictionary<string, object>()
        {
            { "robot_goal_rod_displacement", 0 },
            { "robot_goal_rod_angle", 0 },
            { "robot_2_rod_displacement", 0 },
            { "robot_2_rod_angle", 0 },
            { "robot_5_rod_displacement", 0 },
            { "robot_5_rod_angle", 0 },
            { "robot_3_rod_displacement", 0 },
            { "robot_3_rod_angle", 0 },
        };
        UpdateRods(rods);
    }

    // public void UpdateMessageRandomValues()
    // {
    //     int[] random = new int[8];
    //     Random rand = new Random();
    //     for (int i = 0; i < 8; i++)
    //     {
    //         random[i] = rand.Next(1, 101);
    //     }
    //     Dictionary<string, object> rods = new Dictionary<string, object>()
    //     {
    //         { "robot_goal_rod_displacement", random[0] },
    //         { "robot_goal_rod_angle", random[1] },
    //         { "robot_2_rod_displacement", random[2] },
    //         { "robot_2_rod_angle", random[3] },
    //         { "robot_5_rod_displacement", random[4] },
    //         { "robot_5_rod_angle", random[5] },
    //         { "robot_3_rod_displacement", random[6] },
    //         { "robot_3_rod_angle", random[7] },
    //     };
    //     UpdateRods(rods);
    // }

    public void UpdateRods(Dictionary<string, object> rods)
    {
        foreach (var item in rods)
        {
            data[item.Key] = item.Value;
        }
    }

    private byte[] ConcatBytes(byte[] a, byte[] b)
    {
        byte[] result = new byte[a.Length + b.Length];
        a.CopyTo(result, 0);
        b.CopyTo(result, a.Length);
        return result;
    }
}

File: Assets/Scripts/Python/Message.py
import json 
import numpy as np

class Message:

    def __init__(self, action = ""):
        self.data = {}
        self.data["action"] = action

    
    def encode_to_send(self, length):
        json_data = json.dumps(self.data)
        json_bytes = bytes(json_data,encoding="utf-8")
        if length:           
            return len(json_bytes).to_bytes(4, byteorder="big") + json_bytes
        else:
            return json_bytes


    def decode_from_receive(self, convert):
        self.data = json.loads(convert)

    def request_rods(self):
        rods = {
            "robot_goal_rod_displacement": 0,
            "robot_goal_rod_angle": 0,
            "robot_2_rod_displacement": 0,
            "robot_2_rod_angle": 0,
            "robot_5_rod_displacement": 0,
            "robot_5_rod_angle": 0,
            "robot_3_rod_displacement": 0,
            "robot_3_rod_angle": 0,
        }
        self.data.update(rods)

    def update_message_random_values(self):
        random = np.random.randint(1,101,8)
        rods = {
            "robot_goal_rod_displacement": int(random[0]),
            "robot_goal_rod_angle": int(random[1]),
            "robot_2_rod_displacement": int(random[2]),
            "robot_2_rod_angle": int(random[3]),
            "robot_5_rod_displacement": int(random[4]),
            "robot_5_rod_angle": int(random[5]),
            "robot_3_rod_displacement": int(random[6]),
            "robot_3_rod_angle": int(random[7]),
        }
        self.data.update(rods)   


#test = Message("TEST")
#test.update_message_random_values()
#print(test.encode_to_send())